specs/_features/eip7732/builder.md

```diff
@@ -136,62 +138,86 @@ The builder assembles then
-) -> Sequence[BlobSidecar]:
-    block = signed_block.message
-    block_header = BeaconBlockHeader(
-        slot=block.slot,
-        proposer_index=block.proposer_index,
-        parent_root=block.parent_root,
-        state_root=block.state_root,
-        body_root=hash_tree_root(block.body),
-    )
-    signed_block_header = SignedBeaconBlockHeader(
-        message=block_header, signature=signed_block.signature
-    )
-    sidecars: List[BlobSidecar] = []
-    for index, blob in enumerate(blobs):
-        proof = compute_merkle_proof(
-            blob_kzg_commitments,
-            get_generalized_index(List[KZGCommitment, MAX_BLOB_COMMITMENTS_PER_BLOCK], index),
-        )
-        proof += compute_merkle_proof(
-            block.body,
-            get_generalized_index(
-                BeaconBlockBody,
-                "signed_execution_payload_header",
-                "message",
-                "blob_kzg_commitments_root",
-            ),
-        )
+def get_data_column_sidecars(
+    signed_block_header: SignedBeaconBlockHeader,
+    kzg_commitments: List[KZGCommitment, MAX_BLOB_COMMITMENTS_PER_BLOCK],
+    # [Modified in EIP7732]
+    kzg_commitments_inclusion_proof: Vector[Bytes32, KZG_COMMITMENTS_INCLUSION_PROOF_DEPTH_EIP7732],
+    cells_and_kzg_proofs: Sequence[
+        Tuple[Vector[Cell, CELLS_PER_EXT_BLOB], Vector[KZGProof, CELLS_PER_EXT_BLOB]]
+    ],
+) -> Sequence[DataColumnSidecar]:
+    """
+    Given a signed block header and the commitments, inclusion proof, cells/proofs associated with
+    each blob in the block, assemble the sidecars which can be distributed to peers.
+    """
+    assert len(cells_and_kzg_proofs) == len(kzg_commitments)
+
+    sidecars = []
+    for column_index in range(NUMBER_OF_COLUMNS):
+        column_cells, column_proofs = [], []
+        for cells, proofs in cells_and_kzg_proofs:
+            column_cells.append(cells[column_index])
+            column_proofs.append(proofs[column_index])
         sidecars.append(
-            BlobSidecar(
-                index=index,
-                blob=blob,
-                kzg_commitment=blob_kzg_commitments[index],
-                kzg_proof=blob_kzg_proofs[index],
+            DataColumnSidecar(
+                index=column_index,
+                column=column_cells,
+                kzg_commitments=kzg_commitments,
+                kzg_proofs=column_proofs,
                 signed_block_header=signed_block_header,
-                kzg_commitment_inclusion_proof=proof,
+                kzg_commitments_inclusion_proof=kzg_commitments_inclusion_proof,
             )
         )
     return sidecars
 ```
 
+#### Modified `get_data_column_sidecars_from_block`
+
+*Note*: The function `get_data_column_sidecars_from_block` is modified to
+include the list of blob KZG commitments and to compute the blob KZG commitments
+inclusion proof given that these are in the `ExecutionPayloadEnvelope` now.
+
+```python
+def get_data_column_sidecars_from_block(
```

Comment:
Should we try to abstract this away from the block as much as possible? in principle even the KZG commitment root can be outside of the block and just be in the payload envelope. I think we want to achieve maximum payload+block separation, the only reason we kept the root in the header was to avoid possible off-protocol advantages.

Reply:
Hmm yes I think that would be good. Might want to do this in a separate PR though, as I'm unsure how much more work this would involve.